pkg/securitycontext/admission/admission.go

```diff
@@ -0,0 +1,284 @@
+	}
+
+	// get the namespace
+	namespace, err := c.client.Namespaces().Get(a.GetNamespace())
+	if err != nil {
+		return errors.NewBadRequest(err.Error())
+	}
+
+	// all pods are required to have a service account
+	serviceAccount, err := getServiceAccount(c.client, namespace.Name, pod)
+	if err != nil {
+		return errors.NewBadRequest(err.Error())
+	}
+
+	// get all constraints that are usable by this service account
+	matchedConstraints, err := c.getMatchingSecurityContextConstraints(namespace.Name, serviceAccount)
+	if err != nil {
+		return err
+	}
+
+	providers := make(map[string]scc.SecurityContextConstraintsProvider, 0)
+	// set pre-allocated values on constraints
+	for _, constraint := range matchedConstraints {
+		if requiresPreAllocatedUIDRange(constraint) {
+			min, max, err := getPreallocatedUIDRange(namespace)
+			if err != nil {
+				return err
+			}
+			constraint.RunAsUser.UIDRangeMin = min
+			constraint.RunAsUser.UIDRangeMax = max
+		}
+
+		if requiresPreAllocatedSELinuxLevel(constraint) {
+			// if we don't have options create them here so we can add the preallocated settings
+			if constraint.SELinuxContext.SELinuxOptions == nil {
+				constraint.SELinuxContext.SELinuxOptions = &kapi.SELinuxOptions{}
+			}
+
+			level, err := getPreallocatedLevel(namespace)
+			if err != nil {
+				return err
+			}
+			constraint.SELinuxContext.SELinuxOptions.Level = level
+		}
+
+		provider, err := scc.NewSimpleProvider(constraint, c.client)
+		if err != nil {
+			return err
+		}
+		providers[constraint.Name] = provider
+	}
+
+Containers:
+	for i, container := range pod.Spec.Containers {
+		for providerName, provider := range providers {
+			// Create a security context for this provider
+			context, err := provider.CreateSecurityContext(pod, &container)
+			if err != nil {
+				return err
+			}
+
+			// Validate the copy of the container against the security context provider
+			container.SecurityContext = context
+			errs := provider.ValidateSecurityContext(pod, &container)
+			if len(errs) != 0 {
+				continue
+			}
+
+			// Set the security context on actual container
+			pod.Spec.Containers[i].SecurityContext = context
+			if pod.Annotations == nil {
+				pod.Annotations = make(map[string]string, 0)
+			}
+			pod.Annotations[validatedSCCAnnotation] = providerName
+			continue Containers
+		}
+
+		// If we have reached this code, we couldn't find an SCC that matched
+		// the requested security context for this container
+		//
+		// TODO: better error message
+		return errors.NewBadRequest("Unable to find a valid security context constraint for container")
+	}
+
+	return nil
+}
+
+// getMatchingSecurityContextConstraints returns constraints from the store that match the group,
+// uid, or user of the service account.
+func (c *constraint) getMatchingSecurityContextConstraints(ns string, serviceAccount *kapi.ServiceAccount) ([]*kapi.SecurityContextConstraints, error) {
```

Comment:
make this take a user.Info arg, then it can be reused to get constraints for a service account or the calling user